### ISSUE TITLE
I'm having a hard time understanding logging configuration in 2.0.0

### ISSUE THREAD

**Issue by AdamDotNet** [aspnet/Logging#685]
## Version Info
Windows 10 1703 (15063.540)
VS 2017.3 Preview 7
.Net Core 2.0.0 Preview 2 / 2.0.1-servicing-006924
Microsoft.AspNetCore.All 2.0.0-preview2-final / 2.0.0-rtm-26423

## Question
I see nothing in the Console or Debug-Output windows. My understanding is
```CSharp
WebHost.CreateDefaultBuilder(args)
```
adds those two sinks by default. Why is there no output? Am I doing something wrong? I've tried the NuGet.org preview-2 builds and the MyGet.org daily latest build as of 8/11/2017 (including the SDK/runtime daily build to go with it).

## Code Setup
When I file -> new Asp.Net Core application (Web API) on .Net Core 2.0, I'm not seeing any of my logs show up in the console or the Debug output. The template gives the following config files:

`appsettings.json`
```JSON
{
  "Logging": {
    "IncludeScopes": false,
    "Debug": {
      "LogLevel": {
        "Default": "Warning"
      }
    },
    "Console": {
      "LogLevel": {
        "Default": "Warning"
      }
    }
  }
}
```
`appsettings.Development.json`
```JSON
{
  "Logging": {
    "IncludeScopes": false,
    "LogLevel": {
      "Default": "Debug",
      "System": "Information",
      "Microsoft": "Information"
    }
  }
}
```

My understanding of this config is this:
1. Console and Debug loggers are setup to only log Warnings and more severe logs from appsettings.json.
2. During a development build, global logging is configured so that _all logging providers_ should log Debug and more severe logs, except System and Microsoft are set to Information and more severe; from appsettings.Development.json.
3. Given the default configuration override in appsettings.Development.json, the Console and Debug loggers should write Debug and more severe logs, despite their configuration in appsettings.json.

When I add a logger to ValuesController, like so,
```CSharp
public ValuesController(ILogger<ValuesController> logger)
{
      this.logger = logger;
}
```
and use it in the `Get` method, like so,
```CSharp
logger.LogInformation("The Get method was called.");
```

## Expected Result
1. The information logs are printed to Console and Debug logging providers.
2. The exact same configuration files in Asp.Net Core 1.1.x does print as expected.
3. I should not have to change appsettings.json or appsettings.Development.json to set the Console or Debug specific configurations to log at Debug or more severe when the global logging settings are already configured to do that.

**Comment by AdamDotNet** [comment]
TLDR: It looks like the Console/Debug logging provider only configures from the Console/Debug section, and it ignores the global configuration. Are Console/Debug providers supposed to ignore the global configuration?

**Comment by pakrym** [comment]
The idea is that more specific rules take precedence over less specific. In your case you have the following rules

Logger | Category | MinLevel
--------|------------|-----------
Console | null | Warning
Debug | null | Warning
null | null | Warning
null | System | Information
null | Microsoft| Information


And they are matched the following way:

1. Select all rules that match the provider or its alias. If none are found, select all rules with an empty provider.
1. From the result of the preceding step, select rules with longest matching category prefix. If none are found, select all rules that don't specify a category.
1. If multiple rules are selected take the last one.
1. If no rules are selected, use MinimumLevel.

So for Console/Debug logger Warning would be selected no matter what category is.

Why do you have default levels being provider specific in `appsettings.json`?

More information about it in: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging?tabs=aspnetcore2x "How filtering rules are applied" section

**Comment by AdamDotNet** [comment]
@pakrym Thanks for the detailed response!

Just to reiterate, if a Console, or any provider, is called out specifically in a configuration file, then that setup wins over the global setup. That makes sense, thank you!

I had definitely visited the docs site before asking here. I guess I got hung up on the first part of the article that said, if you have `WebHost.CreateDefaultBuilder(args)`, then expect these logs. But I wasn't getting those results. If I saw "**How filtering rules are applied**" then I might have grasped on sooner.

However,

> Why do you have default levels being provider specific in `appsettings.json`?

The configuration files mentioned in the OP came out that way from a file -> new Asp.Net Core 2.0 template in VS 2017.3 Update 7. If that's a silly setup, I guess the template is out of date. I'm sure if the VS template was right for the new configuration code in 2.0, then it would have worked and I probably would not have asked this question. I just checked the CLI template (`dotnet new WebApi`) too, same `appsettings.json` file.

It's probably safe to close this issue now, as my question was answered. I might have to take a look at the templates repository.

**Comment by pakrym** [comment]
@AdamDotNet just checked templates in latest CLI:
```
{
  "Logging": {
    "IncludeScopes": false,
    "LogLevel": {
      "Default": "Warning"
    }
  }
}

```

```
{
  "Logging": {
    "IncludeScopes": false,
    "LogLevel": {
      "Default": "Debug",
      "System": "Information",
      "Microsoft": "Information"
    }
  }
}
```

**Comment by AdamDotNet** [comment]
@pakrym Hmm, I just got the .Net Core SDK 2.0 RTW and VS 2017.3 RTW (wahoo!) but I'm getting the same templates.

I found this NuGet at
`C:\Program Files\dotnet\sdk\2.0.0\Templates`
`microsoft.dotnet.web.projecttemplates.2.0.1.0.0-beta2-20170803-303.nupkg`

Still has beta terminology in the name. Is that out of date? How do I fix that? Additionally, VS 2017.3 RTW still has a template with the appsettings.json I showed you as well.

**Comment by pakrym** [comment]
@AdamDotNet yep, WebApi template was never fixed.

**Comment by AdamDotNet** [comment]
@pakrym well that explains the discrepancy. Thanks for everything!

**Comment by ian-moore** [comment]
@AdamDotNet thanks for opening this issue and thanks @pakrym for clarifying. This was driving me crazy with a new project not logging out of the ASP.NET Web Api template.

**Comment by springy76** [comment]
>  yep, WebApi template was never fixed.

Please in easy words: What is wrong and how to fix it? I used VS 15.5.x (the latest official non-preview release available at mid of december 2017) and still got non-working shit. Loggers are only outputting warnings and errors.

**Comment by jankalfus** [comment]
The problem with the WebApi template is that it outputs the following for Logging in appsettings.Development.json:

```
    "IncludeScopes": false,
    "LogLevel": {
      "Default": "Debug",
      "System": "Information",
      "Microsoft": "Information"
    }
```

and the following for appsettings.json:

```
"IncludeScopes": false,
    "Debug": {
      "LogLevel": {
        "Default": "Warning"
      }
    },
    "Console": {
      "LogLevel": {
        "Default": "Warning"
      }
    }
```

Since configs get merged and specific rules take over general rules, the rules from appsettings.json take over those in appsettings.Development.json and debug/information level logs are never shown in Development environment.
